python/zfp.pyx

```diff
@@ -0,0 +1,423 @@
+        _set_compression_mode(
+            stream,
+            ztype,
+            ndim,
+            tolerance,
+            rate,
+            precision
+        )
+        mode = zfp_stream_mode(stream)
+        if mode != mode_null and mode != header_mode:
+            raise ValueError(
+                "User-provided zfp_mode {} does not match zfp_mode "
+                "in header {}".format(
+                    mode,
+                    header_mode,
+                )
+            )
+
+    # if the out buffer is a numpy array, check that it's properties match the
+    # header metadata
+    if out is not None and isinstance(out, np.ndarray):
+        # check that numpy and header types match
+        header_dtype = ztype_to_dtype(field[0]._type)
+        if out.dtype != header_dtype:
+            raise ValueError(
+                "Out ndarray has dtype {} but decompression is using "
+                "{}".format(
+                    out.dtype,
+                    header_dtype
+                )
+            )
+
+        # check that numpy and header shape match
+        numpy_shape = [int(x) for x in out.shape[:ndim]]
+        if not all(
+                [x == y for x, y in
+                 zip_longest(numpy_shape, header_shape)
+                ]
+        ):
+            raise ValueError(
+                "Out ndarray has shape {} but decompression is using "
+                "{}".format(
+                    numpy_shape,
+                    header_shape
+                )
+            )
+
+
+cpdef np.ndarray _decompress(
+    bytes compressed_data,
+    out=None,
+    zfp_type ztype=type_none,
+    shape=None,
+    strides=[0,0,0,0],
+    double tolerance = -1,
+    double rate = -1,
+    int precision = -1,
+):
+
+    if compressed_data is None:
+        raise TypeError("compressed_data cannot be None")
+    if compressed_data is out:
+        raise ValueError("Cannot decompress in-place")
+    if shape is not None:
+        _validate_4d_list(shape, "shape")
+    if strides is not None:
+        _validate_4d_list(strides, "strides")
+
+    cdef char* comp_data_pointer = compressed_data
+    cdef zfp_field* field = zfp_field_alloc()
+    cdef bitstream* bstream = stream_open(
+        comp_data_pointer,
+        len(compressed_data)
+    )
+    cdef zfp_stream* stream = zfp_stream_open(bstream)
+    cdef np.ndarray output
+
+    try:
+        zfp_stream_rewind(stream)
+        if zfp_read_header(stream, field, HEADER_FULL) == 0:
+            zfp_stream_rewind(stream)
+            if ztype == type_none or shape is None:
+                raise ValueError(
+                    "Failed to read zfp header and the ztype/shape/mode "
+                    "were not provided"
+                )
+            zshape = gen_padded_int_list(shape, pad=0, length=4)
+            # set the shape, type, and compression mode
+            # strides are set further down
+            field[0].nx, field[0].ny, field[0].nz, field[0].nw = zshape
```

Comment:
I believe these ought to be reversed.

Reply:
This has been resolved by reversing the shape before the padding (line 372 in this preview).